MaxText/train.py

```diff
@@ -267,6 +267,109 @@ def record_activation_metrics(output_metrics, intermediate_outputs, config):
+
+
+def dpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_train=True):
+  """loss_fn for both train and eval.
+
+  Args:
+    model: A nn.Module
+    config: Config of parameters
+    data: Batch of data to apply to the model
+    dropout_rng: A key to use to generate rng for dropout
+    params: Model params
+    is_train: True for train_step and False for eval_step
+
+  Returns:
+    loss: average loss
+    aux: a dictionary including intermediate_outputs, total_loss, and total_weights
+  """
+  # inputs, targets, segments, positions = apply_args
+  rng1, aqt_rng = jax.random.split(dropout_rng)
+
+  # decimate proportion of data when per_device_batch_size<1
+  if is_train:
+    for k, v in data.items():
+      data[k] = v[: config.micro_batch_size_to_train_on, :]
+
+  # for DPO we don't support packed sequence (they shouldn't be present in the first place)
+  data["chosen_segmentation"] = 1 * (data["chosen_segmentation"] == 1)
+  data["rejected_segmentation"] = 1 * (data["rejected_segmentation"] == 1)
+  data["chosen_position"] = data["chosen_position"] * (data["chosen_segmentation"] == 1)
+  data["rejected_position"] = data["rejected_position"] * (data["rejected_segmentation"] == 1)
+
+  inputs = jnp.concatenate([data["chosen"], data["rejected"]], 0)
+  inputs_position = jnp.concatenate([data["chosen_position"], data["rejected_position"]], 0)
+  inputs_segmentation = jnp.concatenate([data["chosen_segmentation"], data["rejected_segmentation"]], 0)
+
+  logits, intermediate_outputs = model.apply(
+      params,
+      inputs,
+      inputs_position,
+      decoder_segment_ids=inputs_segmentation,
+      enable_dropout=config.enable_dropout if is_train else False,
+      rngs={"dropout": rng1, "params": aqt_rng},
+      mutable="intermediates",
+  )
+  ref_logits = model.apply(
+      {"params": reference_params},
+      inputs,
+      inputs_position,
+      decoder_segment_ids=inputs_segmentation,
+      enable_dropout=config.enable_dropout if is_train else False,
+      rngs={"dropout": rng1, "params": aqt_rng},
+  )
+  ref_logits = jax.lax.stop_gradient(ref_logits)
+  n_logits = logits.shape[-3] // 2  # (..., batch, sequence, vocab)
+  chosen_logits, rejected_logits = logits[..., :n_logits, :, :], logits[..., n_logits:, :, :]
+  chosen_ref_logits, rejected_ref_logits = ref_logits[..., :n_logits, :, :], ref_logits[..., n_logits:, :, :]
+
+  chosen_ids = data["chosen"][..., 1:]
+  rejected_ids = data["rejected"][..., 1:]
+  chosen_segmentation = data["chosen_segmentation"][..., 1:]
+  rejected_segmentation = data["rejected_segmentation"][..., 1:]
+
+  chosen_logratios = chosen_logits - chosen_ref_logits
+  chosen_logratios = jnp.take_along_axis(chosen_logratios[..., :-1, :], chosen_ids[..., None], -1)[..., 0]
+  rejected_logratios = rejected_logits - rejected_ref_logits
+  rejected_logratios = jnp.take_along_axis(rejected_logratios[..., :-1, :], rejected_ids[..., None], -1)[..., 0]
+
```

Comment:
While the code employs mathematically equivalent calculations, the DPO loss implementation appears slightly different at first glance.
   - here: chosen_logratios = jnp.take_along_axis(chosen_logits - chosen_ref_logits, chosen_ids)
   - [reference](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L45): chosen_logratios = jnp.take_along_axis(chosen_logits, chosen_ids) - jnp.take_along_axis(chosen_ref_logits, chosen_ref_ids)
  - they are mathematically equivalent but requires some time
 
The separation between `chosen_logratios[..., :-1, :]` and `data["chosen"][..., 1:]` is visually distracting, especially since they're typically grouped together for next-token shifts.

Personally, just curious if it is a better idea to introduce some functions for better modularity and readability only.

Something like:
* get_batch_logps: outputs log_prob after taking in logits and index for both chosen and reject
  - reference: [eric-mitchell dpo](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L90C6-L90C21) & [mlperf dpo](https://github.com/ZhiyuLi-goog/MoE_study/blob/main/hf/dpo_trainers.py#L50)
* dpo_loss: outputs dpo loss after taking in policy_chosen_logps, policy_rejected_logps,ref_chosen_logps, ref_rejected_logps
  - ref: [eric-mitchell](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L45) & [mlperf dpo](https://github.com/ZhiyuLi-goog/MoE_study/blob/main/hf/dpo_trainers.py#L8)

Reply:
Thank you, I was mostly guessing here regarding the correct implementation. I took the gather operation from [https://github.com/huggingface/trl/blob/ed9ea74b62ea5aab3a2c597c4d73fadafc51f5dc/trl/trainer/dpo_trainer.py#L1380](https://github.com/huggingface/trl/blob/ed9ea74b62ea5aab3a2c597c4d73fadafc51f5dc/trl/trainer/dpo_trainer.py#L1380)

`chosen_ref_ids == chosen_ref_ids` - they should be the same right?

I fixed gather (jnp.take_along_axis) to occur twice, it's more readable this way.

When reading DPO reference implementations the helper functions confused me a lot (DPO is quite a simple loss in the end), so I opted for the inlined version, but I'd be happy to modularize them if you think that's better!

Reply:
Thank you @rdyro.

I found a discrepancy between the current implementation and reference one:
* the current implementation: sum(dpo_loss(per_token_logits))
* [reference implementation](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L45-L52): dpo_loss(sum(per_token_logits))
  - note the shape of `*_logps` is (batch_size,) which has already been reduced on vocab dim in [`get_batch_logps`](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L90). In other words dpo_loss applies to per each sequence instead of token, aligning the intuition of DPO about preference between different answers (instead of tokens)

At first glance, they might not be equivalent in math, considering some non-linearity inside dpo_loss i.e. softmax.

Given the above discrepancy, I am also wondering if it is slightly better to modularize it for better understanding.

Reply:
I fully agree with getting the math right, I found anther issue, most implementations add [log_softmax](https://github.com/jcolano/DPO/blob/9ec93f804907179fb66385fc5675fa32a35f22f7/dpo_loss.py#L33) to model outputs before DPO, I added this change.

The discrepancy is there, sorry for that, I fixed it!

On the implementation, I'd really hope to keep the loss in one place, my main difficulty with understanding the DPO loss was how scattered its implementation is across 3 functions, having it in one place should make reading easier. What do you think of this presentation:
```python
# common subsequence and padding mask
common_prefix_mask = jnp.cumsum(chosen_ids != rejected_ids, axis=-1) == 0  # [B, S]
valid_seq_mask = (chosen_segmentation != 0) & (rejected_segmentation != 0) & ~common_prefix_mask  # [B, S]

# compute logratios from the sequence-reduced observed token log-probability
chosen_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(chosen_logits[..., :-1, :], axis=-1), chosen_ids[..., None], axis=-1
)[..., 0]
chosen_logps = jnp.sum(chosen_logps_seq * valid_seq_mask, axis=-1)  # [B]
chosen_ref_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(chosen_ref_logits[..., :-1, :], axis=-1), chosen_ids[..., None], axis=-1
)[..., 0]
chosen_ref_logps = jnp.sum(chosen_ref_logps_seq * valid_seq_mask, axis=-1)  # [B]
chosen_logratios = chosen_logps - chosen_ref_logps  # [B]

rejected_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(rejected_logits[..., :-1, :], axis=-1), rejected_ids[..., None], axis=-1
)[..., 0]
rejected_logps = jnp.sum(rejected_logps_seq * valid_seq_mask, axis=-1)  # [B]
rejected_ref_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(rejected_ref_logits[..., :-1, :], axis=-1), rejected_ids[..., None], axis=-1
)[..., 0]
rejected_ref_logps = jnp.sum(rejected_ref_logps_seq * valid_seq_mask, axis=-1)  # [B]
rejected_logratios = rejected_logps - rejected_ref_logps  # [B]

# DPO loss from chosen and rejected logratios
LABEL_SMOOTHING, BETA = config.dpo_label_smoothing, config.dpo_beta
logratios_delta = BETA * (chosen_logratios - rejected_logratios)  # [B]
loss = (  # [B]
    -jax.nn.log_sigmoid(BETA * logratios_delta) * (1 - LABEL_SMOOTHING)
    - jax.nn.log_sigmoid(-BETA * logratios_delta) * LABEL_SMOOTHING
)
total_loss, total_weights = jnp.sum(loss), jnp.sum(valid_seq_mask)
loss = jnp.sum(loss * (jnp.sum(valid_seq_mask, axis=-1) + 1e-3)) / total_weights
```

Reply:
Sync up and resolved offline!
Thank you @rdyro !